src/lib/parallel_scan/parallel_scan.ml

```diff
@@ -2,925 +2,1579 @@ open Core_kernel
+           f_merge:(   int
+                    -> accum
+                    -> merge_t
+                    -> (accum, final) Continue_or_stop.t M.t)
+        -> f_base:(accum -> base_t -> (accum, final) Continue_or_stop.t M.t)
+        -> init:accum
+        -> finish:(accum -> final M.t)
+        -> (merge_t, base_t) t
+        -> final M.t =
+     fun ~f_merge ~f_base ~init ~finish t ->
+      let open M.Let_syntax in
+      match%bind fold_depth_until' ~f_merge ~f_base ~init t with
+      | Continue result ->
+          finish result
+      | Stop e ->
+          M.return e
+  end
+
+  module Foldable_ident = Make_foldable (Monad.Ident)
+
+  let fold_depth : type merge_t base_t accum.
+         f_merge:(int -> accum -> merge_t -> accum)
+      -> f_base:(accum -> base_t -> accum)
+      -> init:accum
+      -> (merge_t, base_t) t
+      -> accum =
+   fun ~f_merge ~f_base ~init t ->
+    Foldable_ident.fold_depth_until
+      ~f_merge:(fun i acc a -> Continue (f_merge i acc a))
+      ~f_base:(fun acc d -> Continue (f_base acc d))
+      ~init ~finish:Fn.id t
+
+  let fold : type merge_t base_t accum.
+         f_merge:(accum -> merge_t -> accum)
+      -> f_base:(accum -> base_t -> accum)
+      -> init:accum
+      -> (merge_t, base_t) t
+      -> accum =
+   fun ~f_merge ~f_base ~init t ->
+    fold_depth t ~init ~f_merge:(fun _ -> f_merge) ~f_base
+
+  let _fold_until : type merge_t base_t accum final.
+         f_merge:(accum -> merge_t -> (accum, final) Continue_or_stop.t)
+      -> f_base:(accum -> base_t -> (accum, final) Continue_or_stop.t)
+      -> init:accum
+      -> finish:(accum -> final)
+      -> (merge_t, base_t) t
+      -> final =
+   fun ~f_merge ~f_base ~init ~finish t ->
+    Foldable_ident.fold_depth_until
+      ~f_merge:(fun _ -> f_merge)
+      ~f_base ~init ~finish t
+
+  (*List of things that map to a specific level on the tree*)
+  module Data_list = struct
+    module T = struct
+      type 'a t = Single of 'a | Double of ('a * 'a) t [@@deriving sexp]
+    end
+
+    include T
+
+    let rec split : type a. a t -> (a -> a * a) -> (a * a) t =
+     fun lst f ->
+      match lst with
+      | Single a ->
+          Single (f a)
+      | Double t ->
+          let sub = split t (fun (x, y) -> (f x, f y)) in
+          Double sub
+
+    let rec merge : type a. a t -> a t -> (a * a) t =
+     fun lst1 lst2 ->
+      match (lst1, lst2) with
+      | Single a, Single b ->
+          Single (a, b)
+      | Double a, Double b ->
+          Double (merge a b)
+      | _ ->
+          failwith "Cannot merge the two data lists"
+
+    let rec _fold : type a b. a t -> f:(b -> a -> b) -> init:b -> b =
+     fun t ~f ~init ->
+      match t with
+      | Single a ->
+          f init a
+      | Double a ->
+          _fold a ~f:(fun acc (a, b) -> f (f acc a) b) ~init
+
+    (*Just the nested data*)
+    let to_data : type a. a t -> a =
+     fun t ->
+      let rec go : type a. a t -> a * a =
```

Comment:
I think you could rewrite this to be

```ocaml
function
| Single a -> a (* (fst . go) (Single a) = a *)
| Double js -> fst (go js)
```
and then you wouldn't have to call fst on 528. 

Also I would call this function `hd` since that's what it's doing right?